### PR TITLE
5.x - Get tests passing against cake5 branch

### PR DESCRIPTION
**Pull request by markstory**
* Updated dependencies
* Added types from docblocks with phpcbf
* Fixed types until tests pass.

There are a few deprecations that need to be cleaned up but I wanted tests passing first.

### CODE REVIEW

**Comment by markstory** [comment]
Tests are passing now. @othercorey is there anything else I should address before removing usage of QueryTrait?

File: src/Association/Embedded.php [hunk 1]
```diff
@@ -236,12 +238,13 @@ public function setIndexClass($name)
      * Get/set the index class used for this embed.
      *
      * @deprecated 3.2.0 Use setIndexClass()/getIndexClass() instead.
-     * @param string|null|\Cake\ElasticSearch\Index $name The class name to set.
+     * @param \Cake\ElasticSearch\Index|string|null $name The class name to set.
      * @return string The class name.
      */
-    public function indexClass($name = null)
+    public function indexClass(string|Index|null $name = null): string
```
Comment (othercorey) [inline src/Association/Embedded.php:244]:
```suggestion
    public function indexClass(Index|string|null $name = null): string
```

The sniff that converts annotations to type declarations runs before the sniff that fixes the annotation order.

File: src/Datasource/Connection.php [hunk 2]
```diff
@@ -72,7 +74,7 @@ class Connection implements ConnectionInterface
      * @param callable $callback Callback function which can be used to be notified
```
Comment (othercorey) [inline src/Datasource/Connection.php]:
```suggestion
     * @param callable|null $callback Callback function which can be used to be notified
```

File: src/Datasource/Connection.php [hunk 3]
```diff
@@ -330,7 +332,7 @@ public function prepare($sql)
      * @see \Cake\Datasource\ConnectionInterface::getDriver()
      * @return \Elastica\Client
      */
-    public function getDriver()
+    public function getDriver(): object
```
Comment (othercorey) [inline src/Datasource/Connection.php]:
```suggestion
    public function getDriver(): Client
```

This should be able to narrow the return type.

File: src/Datasource/Connection.php [hunk 4]
```diff
@@ -360,11 +362,11 @@ public function newQuery()
     /**
      * Returns the index for the given connection
      *
-     * @param  string $name Index name to create connection to, if no value is passed
+     * @param string $name Index name to create connection to, if no value is passed
```
Comment (othercorey) [inline src/Datasource/Connection.php]:
```suggestion
     * @param string|null $name Index name to create connection to, if no value is passed
```

File: src/Document.php [hunk 5]
```diff
@@ -34,18 +34,18 @@ class Document implements EntityInterface
...
      * a document representing an entity in a elastic search index,
      *
-     * @param array|\Elastica\Result $data An array or Result object that
+     * @param \Elastica\Result|array $data An array or Result object that
      *  represents an Elasticsearch document
      * @param array $options An array of options to set the state of the
      *  document
      */
-    public function __construct($data = [], $options = [])
+    public function __construct(array|Result $data = [], array $options = [])
```
Comment (othercorey) [inline src/Document.php]:
```suggestion
    public function __construct(Result|array $data = [], array $options = [])
```

File: src/Index.php [hunk 6]
```diff
@@ -856,13 +860,13 @@ public function patchEntity(EntityInterface $entity, array $data, array $options
...
      *
-     * @param array|\Traversable $entities the entities that will get the
+     * @param \Traversable|array $entities the entities that will get the
      * data merged in
      * @param array $data list of arrays to be merged into the entities
      * @param array $options A list of options for the objects hydration.
      * @return array
      */
-    public function patchEntities($entities, array $data, array $options = []): array
+    public function patchEntities(array|Traversable $entities, array $data, array $options = []): array
```
Comment (othercorey) [inline src/Index.php]:
```suggestion
    public function patchEntities(Traversable|array $entities, array $data, array $options = []): array
```

File: src/Marshaller.php [hunk 7]
```diff
@@ -136,9 +141,9 @@ protected function createAndHydrate($class, array $data, array $options = [], $i
      * @param \Cake\ElasticSearch\Association\Embedded $embed   The embed definition.
      * @param array                                    $data    The data to marshal
      * @param array                                    $options The options to pass on
-     * @return array|\Cake\ElasticSearch\Document Either a document or an array of documents.
+     * @return \Cake\ElasticSearch\Document|array Either a document or an array of documents.
      */
-    protected function newNested(Embedded $embed, array $data, array $options = [])
+    protected function newNested(Embedded $embed, array $data, array $options = []): array|Document
```
Comment (othercorey) [inline src/Marshaller.php]:
```suggestion
    protected function newNested(Embedded $embed, array $data, array $options = []): Document|array
```

File: src/Marshaller.php [hunk 8]
```diff
@@ -163,9 +168,9 @@ protected function newNested(Embedded $embed, array $data, array $options = [])
      * @param \Cake\ElasticSearch\Association\Embedded $embed The embed definition.
      * @param \Cake\ElasticSearch\Document|array $existing The existing entity or entities.
      * @param array $data The data to marshal
-     * @return array|\Cake\ElasticSearch\Document Either a document or an array of documents.
+     * @return \Cake\ElasticSearch\Document|array Either a document or an array of documents.
      */
-    protected function mergeNested(Embedded $embed, $existing, array $data)
+    protected function mergeNested(Embedded $embed, Document|array|null $existing, array $data): array|Document
```
Comment (othercorey) [inline src/Marshaller.php]:
```suggestion
    protected function mergeNested(Embedded $embed, Document|array|null $existing, array $data): Document|array
```

File: src/Query.php [hunk 9]
```diff
@@ -115,7 +117,7 @@ public function __construct(Index $repository)
      * @param bool $overwrite Whether or not to replace previous selections.
      * @return $this
      */
-    public function select($fields, bool $overwrite = false)
+    public function select(Closure|array|string|int|float $fields, bool $overwrite = false)
```
Comment (othercorey) [inline src/Query.php:120]: Need to update the docblock

File: src/Query.php [hunk 10]
```diff
@@ -146,7 +148,7 @@ public function limit($limit)
      * @param int $num The number of records to be skipped
      * @return $this
      */
-    public function offset($num)
+    public function offset(?int $num)
```
Comment (othercorey) [inline src/Query.php:151]: Need to update docblock if null is supported.

File: src/Query.php [hunk 11]
```diff
@@ -132,7 +134,7 @@ public function select($fields, bool $overwrite = false)
      * @param int $limit The number of documents to return.
      * @return $this
      */
-    public function limit($limit)
+    public function limit(?int $limit)
```
Comment (othercorey) [inline src/Query.php:137]: Need to update docblock if null is supported.

File: src/Query.php [hunk 12]
```diff
@@ -221,11 +223,11 @@ public function clause($name)
      * - ['name' => 'asc', 'price' => 'desc']
      * - 'field1' (defaults to order => 'desc')
      *
-     * @param string|array $order The sorting order to use.
+     * @param array|string $order The sorting order to use.
      * @param bool $overwrite Whether or not to replace previous sorting.
      * @return $this
      */
-    public function order($order, $overwrite = false)
+    public function order(Closure|array|string $order, bool $overwrite = false)
```
Comment (othercorey) [inline src/Query.php:230]: Need to update docblock

File: src/Query.php [hunk 13]
```diff
@@ -268,9 +270,9 @@ public function order($order, $overwrite = false)
      *
      * @param string $finder The finder method to use.
      * @param array $options The options for the finder.
-     * @return \Cake\ElasticSearch\Query
+     * @return self
```
Comment (othercorey) [inline src/Query.php]:
```suggestion
     * @return static
```

File: src/Query.php [hunk 14]
```diff
@@ -315,14 +317,17 @@ public function find($finder = 'all', array $options = [])
      *   $query->where(new \Elastica\Filter\Term('name.first', 'jose'));
      * }}{
      *
-     * @param array|null|callable|\Elastica\Query\AbstractQuery $conditions The list of conditions.
+     * @param \Elastica\Query\AbstractQuery|callable|array|null $conditions The list of conditions.
```
Comment (othercorey) [inline src/Query.php]: Should this be \Closure not callable?

File: src/Query.php [hunk 15]
```diff
@@ -452,10 +457,10 @@ public function setFullQuery(AbstractQuery $query)
     /**
      * Add collapse to the elastic query object
      *
-     * @param string|\Elastica\Collapse $collapse Collapse field or elastic collapse object
+     * @param \Elastica\Collapse|string $collapse Collapse field or elastic collapse object
      * @return $this
      */
-    public function collapse($collapse)
+    public function collapse(string|Collapse $collapse)
```
Comment (othercorey) [inline src/Query.php]:
```suggestion
    public function collapse(Collapse|string $collapse)
```

File: src/Query.php [hunk 16]
```diff
@@ -469,10 +474,10 @@ public function collapse($collapse)
     /**
      * Add an aggregation to the elastic query object
      *
-     * @param  array|\Elastica\Aggregation\AbstractAggregation $aggregation One or multiple facets
+     * @param \Elastica\Aggregation\AbstractAggregation|array $aggregation One or multiple facets
      * @return $this
      */
-    public function aggregate($aggregation)
+    public function aggregate(array|AbstractAggregation $aggregation)
```
Comment (othercorey) [inline src/Query.php]:
```suggestion
    public function aggregate(AbstractAggregation|array $aggregation)
```

File: src/Query.php [hunk 17]
```diff
@@ -507,14 +512,18 @@ public function searchOptions(?array $options = null)
...
      * @param bool $overwrite Whether or not to replace previous query.
      * @param string $type The method to use for appending the conditions to the Query
      * @return $this
      */
-    protected function _buildBoolQuery($partType, $conditions, $overwrite, $type = 'addMust')
-    {
-        if ($this->_queryParts[$partType] === null || $overwrite) {
+    protected function _buildBoolQuery(
+        string $partType,
+        array|callable|AbstractQuery $conditions,
```
Comment (othercorey) [inline src/Query.php]:
```suggestion
        AbstractQuery|callable|array $conditions,
```

File: src/QueryBuilder.php [hunk 18]
```diff
@@ -216,25 +246,25 @@ public function gte($field, $value)
...
      * documents that have child docs matching the query.
      *
-     * @param string|\Elastica\Query|\Elastica\Query\AbstractQuery $query The query.
+     * @param \Elastica\Query|\Elastica\Query\AbstractQuery|string $query The query.
      * @param string $type The child type to query against.
      * @return \Elastica\Query\HasChild
      * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-has-child-query.html
      */
-    public function hasChild($query, $type)
+    public function hasChild(string|Query|AbstractQuery $query, string $type): HasChild
```
Comment (othercorey) [inline src/QueryBuilder.php]:
```suggestion
    public function hasChild(Query|AbstractQuery|string $query, string $type): HasChild
```

File: src/QueryBuilder.php [hunk 19]
```diff
@@ -216,25 +246,25 @@ public function gte($field, $value)
...
      * Query by child documents having parent documents matching the query
      *
-     * @param string|\Elastica\Query|\Elastica\Query\AbstractQuery $query The query.
+     * @param \Elastica\Query|\Elastica\Query\AbstractQuery|string $query The query.
      * @param string $type The parent type to query against.
      * @return \Elastica\Query\HasParent
      * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-has-parent-query.html
      */
-    public function hasParent($query, $type)
+    public function hasParent(string|Query|AbstractQuery $query, string $type): HasParent
```
Comment (othercorey) [inline src/QueryBuilder.php]:
```suggestion
    public function hasParent(Query|AbstractQuery|string $query, string $type): HasParent
```

File: src/QueryBuilder.php [hunk 20]
```diff
@@ -409,11 +439,11 @@ public function regexp($field, $regexp, $boost = 1.0)
      *  $builder->script("doc['price'].value > 1");
      * }}}
      *
-     * @param array|string|\Elastica\Script\AbstractScript $script The script.
+     * @param \Elastica\Script\AbstractScript|array|string $script The script.
      * @return \Elastica\Query\Script
      * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-script-query.html
      */
-    public function script($script)
+    public function script(array|string|AbstractScript $script): Script
```
Comment (othercorey) [inline src/QueryBuilder.php]:
```suggestion
    public function script(AbstractScript|array|string $script): Script
```

**Comment by othercorey** [comment]
> Tests are passing now. @othercorey is there anything else I should address before removing usage of QueryTrait?

Just some type cleanup due to code sniff limitations.

**Pull request merged**